// @open-draft/deferred-promise/build/DeferredPromise.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeferredPromise = void 0;
const createDeferredExecutor_1 = require("./createDeferredExecutor");
class DeferredPromise extends Promise {
    #executor;
    resolve;
    reject;
    constructor(executor = null) {
        const deferredExecutor = (0, createDeferredExecutor_1.createDeferredExecutor)();
        super((originalResolve, originalReject) => {
            deferredExecutor(originalResolve, originalReject);
            executor?.(deferredExecutor.resolve, deferredExecutor.reject);
        });
        this.#executor = deferredExecutor;
        this.resolve = this.#executor.resolve;
        this.reject = this.#executor.reject;
    }
    get state() {
        return this.#executor.state;
    }
    get rejectionReason() {
        return this.#executor.rejectionReason;
    }
    then(onFulfilled, onRejected) {
        return this.#decorate(super.then(onFulfilled, onRejected));
    }
    catch(onRejected) {
        return this.#decorate(super.catch(onRejected));
    }
    finally(onfinally) {
        return this.#decorate(super.finally(onfinally));
    }
    #decorate(promise) {
        return Object.defineProperties(promise, {
            resolve: { configurable: true, value: this.resolve },
            reject: { configurable: true, value: this.reject },
        });
    }
}
exports.DeferredPromise = DeferredPromise;
